### ISSUE TITLE
Potential LJcos

### ISSUE THREAD

**Issue by niktre** [espressopp/espressopp#31]
Hi,

I would like to use the potential LJcos and there is some parameter there which I do not understand. In LJcos.hpp there is the following explanation:
real auxCoef;  // This is temporary solution for empty potential. This problem
                       // appears when there is a potential for, for example,  00 and
                       // 01, but no potential for 11. The loop in interaction
                       // template will call potential(1, 1) anyway if these particles are
                       // within the range. Thus one have to set coefficient to 0,
                       // in order to get zero forces.

what are "00", "01" and "11"? Is this still up-to-date?

**Comment by MrTheodor** [comment]
Hi, 
Those fancy "00", "01".. are the shortcuts for pair types 0-0, 0-1 ;-)
AFAIK this work in the way that if there is no pair tuple in the potential array defined then it will be added anyway with potential object constructed without any parameters (LJCos.hpp:57). Therefore in such case the _computeForce, _computeEnergy should not do anything.

**Comment by acfogarty** [comment]
I would guess this refers to a system where there are two atom types, called 0 and 1. You might want to set an interaction between types 0 and 0 and between types 0 and 1, but not between 1 and 1.

Here's the relevant bit of the interaction template:
```
int type1 = p1.type();
int type2 = p2.type();
const Potential &potential = getPotential(type1, type2);
Real3D force(0.0);
if (potential._computeForce(force, p1, p2)) {
          p1.force() += force;
          p2.force() -= force;
}
```
`getPotential` returns `potentialArray.at(type1, type2)` which I guess throws an exception if potentialArray doesn't have an entry for type1,type2

**Comment by vitst** [comment]
Jakub is right. I do not remember exactly, but at some condition it will initiate an error.

**Comment by acfogarty** [comment]
So here's my guess:

potentialArray is of type esutil::Array2D, initialised to size (0,0)

When you call setPotential(type1,type2), potentialArray is resized if necessary to be as big as size (type1,type2) and any missing elements are filled in with the protype that was specified when potentialArray was declared, the protype is Potential(), see e.g. VerletListInteractionTemplate.hpp:389

So then later, in addForces in the loop over all pairs (i.e. in whichever interaction template is being used)
```
const Potential &potential = getPotential(type1, type2);
```
returns this empty potential if there's no user-defined entry in potentialArray for type1,type2, and then
```
potential._computeForce(force, p1, p2)
```
either calculates force=0 or maybe in some cases returns 0?

**Comment by MrTheodor** [comment]
@acfogarty even more, in addForces, when potential is not found, the array will also be resized. This is because set and get works with at method of Array2D.
```
 When the element doesn't exist, the Array2D is automatically
	  extended so that the element exists. The new elements are
	  initialised with copies of the prototype.
```

**Comment by niktre** [comment]
Ok, thanks for the clarification. Closed #31
